mlflow/pyfunc/__init__.py

```diff
@@ -656,15 +674,285 @@ def _warn_dependency_requirement_mismatches(model_path):
+        )
+        if not os.path.exists(self._python_installation_dir):
+            # install specified python version
+            sub_env = os.environ.copy()
+            sub_env['HOME'] = pyenv_home_path
+            subprocess.run(
+                [_PYENV_CMD, 'install', '-v', self._py_version],
+                check=True, stdout=sys.stdout, stderr=sys.stderr,
+                env=sub_env
+            )
+        else:
+            _logger.info(f'Found python installation cache at {self._python_installation_dir}')
+        python_bin_path = os.path.join(
+            self._python_installation_dir, 'bin/python'
+        )
+        req_file_content_sha = get_file_sha_hexdigest(self._requirements_file_path)
+        self._virtual_env_dir = os.path.join(
+            env_cache_dir, 'venvs', req_file_content_sha
+        )
+        if not os.path.exists(self._virtual_env_dir):
+            # install virtual env
+            subprocess.run(
+                [_VIRTUALENV_CMD, f'--python={python_bin_path}', self._virtual_env_dir]
+            )
+            virtual_env_pip = os.path.join(
+                self._virtual_env_dir, 'bin/pip'
+            )
+            subprocess.run(
+                [virtual_env_pip, 'install', '-r', self._requirements_file_path],
+                check=True, stdout=sys.stdout, stderr=sys.stderr
+            )
+            # For debug purpose, install the dev mlflow version in virtual env
+            subprocess.run(
+                [virtual_env_pip, 'install', '-e',
+                 os.path.dirname(os.path.dirname(mlflow.__file__))]
+            )
+        else:
+            _logger.info(f'Found python virtual env cache at {self._virtual_env_dir}')
+
+        self._virtual_env_python_bin = os.path.join(
+            self._virtual_env_dir, 'bin/python'
+        )
+
+    def create_subproc(self):
+        """
+        This method can be called from driver side or executor side.
+        """
+        # setup tcp connection with subproc
+        listen_sock = socket.socket()
+        try:
+            listen_sock.bind(('127.0.0.1', 0))
+            server_port = listen_sock.getsockname()[1]
+            listen_sock.settimeout(10)
+            listen_sock.listen(1)
+
+            self._subproc = subprocess.Popen(
+                [self._virtual_env_python_bin,
+                 '-m', 'mlflow.pyfunc.restored_python_env_model_infer_subproc',
+                 '--server_port', str(server_port),
+                 '--loader_module', self._loader_module,
+                 '--model_path', self._model_path,
+                 ],
+                stdout=sys.stdout, stderr=sys.stderr,
+            )
+
+            comm_sock, _ = listen_sock.accept()
+            self._comm_sock = comm_sock
+            self._comm_stream = comm_sock.makefile('rwb')
+        except socket.timeout:
+            raise RuntimeError('Launch sub process for virtual python environment failed.')
```

Comment:
Are we going to have an ability to capture this exception and run with current system environment if something goes wrong with the environment re-build (i.e. some OSS packages delisting a particular version of a package due to bugs)?

Reply:
> i.e. some OSS packages delisting

Is this possible ? I think version published to pypi cannot be withdrawn.

And if subprocess failed due to any reason, I would like to let the error raised and let user determine what to do next.